akka-actor/src/main/scala/akka/util/WildcardIndex.scala

```diff
@@ -0,0 +1,74 @@
+/**
+ * Copyright (C) 2009-2016 Lightbend Inc. <http://www.lightbend.com>
+ */
+
+package akka.util
+
+import scala.annotation.tailrec
+import scala.collection.immutable.HashMap
+
+private[akka] final case class WildcardIndex[T](wildcardTree: WildcardTree[T] = WildcardTree[T](), doubleWildcardTree: WildcardTree[T] = WildcardTree[T]()) {
+
+  val empty = WildcardTree[T]()
+
+  def insert(elems: Array[String], d: T): WildcardIndex[T] = elems.lastOption match {
+    case Some("**") ⇒ copy(doubleWildcardTree = doubleWildcardTree.insert(elems.iterator, d))
+    case Some(_)    ⇒ copy(wildcardTree = wildcardTree.insert(elems.iterator, d))
+    case _          ⇒ this
+  }
+
+  def forNonEmpty(tree: WildcardTree[T])(f: WildcardTree[T] ⇒ WildcardTree[T]) = tree match {
+    case `empty` ⇒ empty
+    case _       ⇒ f(tree)
+  }
+
+  def find(elems: Iterable[String]): Option[T] = {
+    (forNonEmpty(wildcardTree)(_.findWithSingleWildcard(elems.iterator)) match {
+      case `empty` ⇒ forNonEmpty(doubleWildcardTree)(_.findWithTerminalDoubleWildcard(elems.iterator))
```

Comment:
that case (and the one in `forNonEmpty`)  will be execute `equals` on the `WildcardTree`, which might be too costly.
Introduce an `isEmpty` method on the `WildcardTree` and use `if` statements.

I could also nitpick on that using `forNonEmpty` allocates the `f` thunk. Optimal performance would be to write the hole thing inline with if statements.

Yeah, performance is more important than readability here.